src/mongocxx/test/collection.cpp

```diff
@@ -2721,4 +2721,53 @@ TEST_CASE("find_and_x operations append write concern correctly", "[collection]"
     REQUIRE(called);
 }
 
+TEST_CASE("Ensure that the WriteConcernError 'errInfo' object is propagated", "[collection]") {
+    using namespace bsoncxx;
+    instance::current();
+
+    client mongodb_client{uri{}};
+
+    using bsoncxx::builder::basic::sub_document;
+    auto err_info = builder::basic::document{};
+    err_info.append(kvp("writeConcern", [](sub_document sub_doc) {
+        sub_doc.append(kvp("w", types::b_int32{2}));
+        sub_doc.append(kvp("wtimeout", types::b_int32{0}));
+        sub_doc.append(kvp("provenance", "clientSupplied"));
+    }));
+
+    auto fail_point = builder::basic::document{};
+    fail_point.append(kvp("configureFailPoint", "failCommand"));
+
+    using bsoncxx::builder::basic::sub_array;
+    fail_point.append(kvp("data", [&err_info](sub_document sub_doc) {
+        sub_doc.append(
+            kvp("failCommands", [&err_info](sub_array sub_arr) { sub_arr.append("insert"); }));
+        sub_doc.append(kvp("writeConcernError", [&err_info](sub_document sub_doc) {
+            sub_doc.append(kvp("code", types::b_int32{100}));
+            sub_doc.append(kvp("codeName", "UnsatisfiableWriteConcern"));
+            sub_doc.append(kvp("errmsg", "Not enough data-bearing nodes"));
+            sub_doc.append(kvp("errInfo", types::b_document{err_info}));
+        }));
+    }));
+
+    fail_point.append(
+        kvp("mode", [](sub_document sub_doc) { sub_doc.append(kvp("times", types::b_int32{1})); }));
+
+    mongodb_client["admin"].run_command(fail_point.view());
+    collection coll = mongodb_client["test"]["errInfo"];
+
+    coll.drop();
+    auto doc = make_document(kvp("x", types::b_int32{1}));
+
+    bsoncxx::document::element result;
+    try {
+        coll.insert_one(doc.view());
+    } catch (const operation_exception& e) {
+        auto error = e.raw_server_error()->view();
+        result = error["writeConcernErrors"][0]["errInfo"];
```

Comment:
I'm not 100% sure, but the cause of the failing Valgrind tasks may be that once `e` goes out of scope, the view to `errInfo` is no longer valid. Perhaps copying out "errInfo" to a `bsoncxx::document::value` will fix that.

FWIW tests passed on the last commit, so I suspect it's this. I don't see why the other changes would be causing that.

Reply:
I J\just did the testing inside the catch to avoid having to copy out err info.

Reply:
Oh nice, even simpler.